solr/core/src/java/org/apache/solr/core/DirectoryFactory.java

```diff
@@ -166,6 +166,18 @@ public long size(Directory directory) throws IOException {
     return sizeOfDirectory(directory);
   }
 
+  /**
+   * @param directory to calculate size of
+   * @return size in bytes on disk, regardless of compression, etc
+   * @throws IOException on low level IO error
+   */
+  public long onDiskSize(Directory directory) throws IOException {
+    if (directory instanceof OnDiskSizeDirectory) {
+      return onDiskSizeOfDirectory(directory);
+    }
+    return sizeOfDirectory(directory);
```

Comment:
Note: at the `Directory` level it totally makes sense to fallback to `sizeOfDirectory()` here. This is a bit counterintuitive considering that falling back was the source of the "drift" bug, but it's different at the `Directory` level:

Any Directory impl that makes no distinction between "onDisk" and "logical" does so because it's writing everything to disk. It also makes sense to continue to refer to this as "onDisk" as opposed to "compressed" or whatever, because compression is orthogonal. Also, the fact that the access copy in our case is also technically "on disk" -- albeit an ephemeral local ssd -- is an implementation detail. For semantic purposes the access copy should be considered a cache.